README.md
# Autoprotocol Python Library

[![Build Status](https://travis-ci.org/autoprotocol/autoprotocol-python.svg?branch=master)](https://travis-ci.org/autoprotocol/autoprotocol-python)
[![PyPI version](https://img.shields.io/pypi/v/autoprotocol.svg?maxAge=2592000)](https://pypi.python.org/pypi/autoprotocol)

### **[View Library Documentation on readthedocs.org](http://autoprotocol-python.readthedocs.org/en/latest/)**

[Autoprotocol](http://www.autoprotocol.org) is a standard way to express
experiments in life science. This repository contains a python library for
generating Autoprotocol.

## Installation

    $ git clone https://github.com/autoprotocol/autoprotocol-python
    $ cd autoprotocol-python
    $ python setup.py install

or, alternatively:

    $ pip install autoprotocol

**check the releases tab or the [changelog](http://autoprotocol-python.readthedocs.io/en/latest/changelog.html) in this repository to see the latest release that will be downloaded.  To be completely up to date it's safest to clone and install this repository as above**

## Building a Protocol
A basic protocol object has empty "refs" and "instructions" stanzas.  Various helper methods in the Protocol class are used to append Instructions and Refs to the Protocol object such as in the simple protocol below:

```python
import json
from autoprotocol.protocol import Protocol

#instantiate new Protocol object
p = Protocol()

# append refs (containers) to Protocol object
bacteria = p.ref("bacteria", cont_type="96-pcr", storage="cold_4")
media = p.ref("media", cont_type="micro-1.5", storage="cold_4")
reaction_plate = p.ref("reaction_plate", cont_type="96-flat", storage="warm_37")

# distribute media from 1.5mL tube to reaction wells
p.distribute(media.well(0).set_volume("1000:microliter"),
             reaction_plate.wells_from(0,4), ["140:microliter",
             "130:microliter", "120:microliter", "100:microliter"])

# transfer bacteria from source wells to reaction wells
p.transfer(bacteria.wells_from(0,4), reaction_plate.wells_from(0,4),
           ["10:microliter", "20:microliter", "30:microliter", "40:microliter"])

# cover plate
p.cover(reaction_plate)

# incubate bacteria at 37 degrees for 5 hours
p.incubate(reaction_plate, "warm_37", "5:hour", shaking=True)

# read absorbance of the first four wells on the reaction plate at 600 nanometers
p.absorbance(reaction_plate, reaction_plate.wells_from(0,4).indices(), "600:nanometer",
             "OD600_reading_01092014")

print json.dumps(p.as_dict(), indent=2)
```
The script above prints the following output to standard out by calling as_dict() on the Protocol object:

```
{
  "refs": {
    "media": {
      "new": "micro-1.5",
      "store": {
        "where": "cold_4"
      }
    },
    "bacteria": {
      "new": "96-pcr",
      "store": {
        "where": "cold_4"
      }
    },
    "reaction_plate": {
      "new": "96-flat",
      "store": {
        "where": "warm_37"
      }
    }
  },
  "instructions": [
    {
      "groups": [
        {
          "distribute": {
            "to": [
              {
                "volume": "140.0:microliter",
                "well": "reaction_plate/0"
              },
              {
                "volume": "130.0:microliter",
                "well": "reaction_plate/1"
              },
              {
                "volume": "120.0:microliter",
                "well": "reaction_plate/2"
              },
              {
                "volume": "100.0:microliter",
                "well": "reaction_plate/3"
              }
            ],
            "from": "media/0",
            "allow_carryover": false
          }
        },
        {
          "transfer": [
            {
              "volume": "10.0:microliter",
              "to": "reaction_plate/0",
              "from": "bacteria/0"
            }
          ]
        },
        {
          "transfer": [
            {
              "volume": "20.0:microliter",
              "to": "reaction_plate/1",
              "from": "bacteria/1"
            }
          ]
        },
        {
          "transfer": [
            {
              "volume": "30.0:microliter",
              "to": "reaction_plate/2",
              "from": "bacteria/2"
            }
          ]
        },
        {
          "transfer": [
            {
              "volume": "40.0:microliter",
              "to": "reaction_plate/3",
              "from": "bacteria/3"
            }
          ]
        }
      ],
      "op": "pipette"
    },
    {
      "lid": "standard",
      "object": "reaction_plate",
      "op": "cover"
    },
    {
      "where": "warm_37",
      "object": "reaction_plate",
      "co2_percent": 0,
      "duration": "5:hour",
      "shaking": true,
      "op": "incubate"
    },
    {
      "dataref": "OD600_reading_01092014",
      "object": "reaction_plate",
      "wells": [
        "A1",
        "A2",
        "A3",
        "A4"
      ],
      "num_flashes": 25,
      "wavelength": "600:nanometer",
      "op": "absorbance"
    }
  ]
}
```
## Extras

A folder of SublimeText snippets for this library is included in this repo.  To use them, copy the folder to `~/Library/Application\ Support/Sublime\ Text\ 3/Packages/User`
(replace with the version of Sublime Text you're using if it's not 3).

## Disclaimers

The Autoprotocol Python Library contains official support for units listed in the [Autoprotocol](http://www.autoprotocol.org) specification. Other units and abbreviations of units have been provided as a convenience.

## Contributing

The easiest way to contribute is to fork this repository and submit a pull
request.  You can also submit an issue or write an email to us at
support@transcriptic.com if you want to discuss ideas or bugs.

autoprotocol-python is BSD licensed (see [LICENSE](http://autoprotocol-python.readthedocs.io/en/latest/LICENSE.html)).
Before we can accept your pull request, we require that you sign a CLA (Contributor License Agreement)
allowing us to distribute your work under the BSD license. Email one of the [AUTHORS](http://autoprotocol-python.readthedocs.io/en/latest/AUTHORS.html) for more details.


CHANGELOG.md
# CHANGELOG

This project adheres to [Semantic Versioning](http://semver.org/)

## Unreleased
---
Added

Changed

Removed

Fixed


## v3.7.6 - 2016-07-25
---
Added

Changed
- documentation
- convert pipette operations to microliters

Removed

Fixed
- cover_types on 96-deep-kf and 96-deep
- dispense_speed and distribute_target in Distribute instruction

## v3.7.5 - 2016-07-08
---
Added
- plate type 6-flat-tc to ContainerType

Changed

Removed

Fixed


## v3.7.4 - 2016-07-07
---
Added

Changed

Removed

Fixed
- auto-uncover before provision instructions

## v3.7.3 - 2016-07-06
---
Added
- is_resource_id added to dispense and dispense_full_plate instructions
- dispense instruction tests
- autocover before incubate
- assertions and tests for flow_analyze
- WellGroup methods: set_group_name, pop, insert, wells_with
- documentation

Changed
- WellGroup.extend(wells) can now take in a list of wells
- dispense instruction json outputs
- removed capability 'cover' from 96-pcr and 384-pcr plates

Removed

Fixed
- spin auto-cover
- compatibility with py3 in flow_analyze


## v3.7.2 - 2016-06-24
---
Added
- validations before implicit cover or seal
- new plate types 384-clear-clear, 384-flat-white-white-lv, 384-flat-white-white-tc

Changed

Removed

Fixed

## v3.7.1 - 2016-06-17
---
Added
- validations of input types before cover check
- cover_types and seal_types to _CONTAINER_TYPES

Changed

Removed
- string input types for source, destination wells for Instructions Consolidate, Autopick, Mix

Fixed

## v3.7.0 - 2016-06-14
---
Added
- track plate cover status - Container objects now have a `cover` attribute, implicit plate unsealing or uncovering prior to steps that require the plate to be uncovered.

Changed

Removed

Fixed
- Stamp separates row stamps with more than 2 containers

## v3.6.0 - 2016-06-07
---
Added
- Time constraints

Changed
- illuminaseq allows cycle specification

Removed

Fixed

## v3.5.3 - 2016-05-16
---
Added

Changed
- harness.py returns proper boolean for thermocycle types

Removed

Fixed

## v3.5.2 - 2016-05-13
---
Added
- Unit specific error handling

Changed

Removed

Fixed
- thermocycle gradient steps in harness.py

## v3.5.1 - 2016-05-12
---
Added
- mix allows one_tip=True

Changed

Removed

Fixed
- acoustic_transfer handling of droplet size

## v3.5.0 - 2016-05-06
---
Added
- measure_mass instruction
- measure_volume instruction
- illumina_sequence instruction


Changed
- gel purify parameters have changed
- Spin instruction takes directional parameters

Removed

Fixed
 - WellGroup checks that all elements are wells
 - Concatenation of Well to WellGroup no longer returns None
 - gel string in documentation
 - fix harness to be python3 compatible
 - Compatibility of Unit for acceleration

## v3.4.0 - 2016-04-22
---
Added
 - discard() and and set_storage() methods for containers
 - gel_purify instruction to instruction.py and protocol.py
 - support for list input type for humanize and robotize (container and container_type)

Changed

Removed

Fixed

## v3.3.0 - 2016-04-13
---
Added
- csv-table input type to harness.py

Changed

Removed

Fixed

## v3.2.0 - 2016-04-07
---
Added
- additional parameter, gain, to fluorescence
- checking for valid plate read incubate parameters
- Unit(Unit(...)) now returns a Unit
- disclaimer to README.md on unit support
- Unit support for `molar`

Changed
- adding magnetic transfer functions to documentation
- magnetic transfer instructions to now pass relevant inputs through units

Removed

Fixed
- documentation for magnetic transfer instructions correctly uses hertz

## v3.1.0 - 2016-03-24
---
Added
- additional parameters to spectrophotometry instructions (absorbance, luminescence, fluorescence) to instruction.py and protocol.py
- helper function in util.py to create incubation dictionaries
- support for a new instruction for measure_concentration

Changed
- Updated handling of multiplication and division of Units of the same dimension to automatically resolve when possible

Removed

Fixed
- Updated maximum tip capacity for a transfer operation to 900uL instead of 750uL
- Updated Unit package to default to `Autoprotocol` format representation for temperature and speed units


## v3.0.0 - 2016-03-17
---
Added
- container+ input type to harness.py
- magnetic_transfer instruction to instruction.py and protocol.py
- kf container types in container_type.py

Changed
- release versioning has been removed in favor of protocol versioniong in harness.py
- update 6-flat well volumes
- Unit now uses Pint's Quantity as a base class

Removed

Fixed
- default versioning in manifest_test.json
- Update container_test.py and container_type_test.py to include safe_min_volume_ul

## v2.7.0 - 2016-02-18
---
Added
- safe_min_volume_ul in _CONTAINER_TYPES

Changed
- updated dead_volume_ul values in _CONTAINER_TYPES

Removed

Fixed
- Protocol.stamp() smartly calculates max_tip_volume using residual volumes

## v2.6.0 - 2015-02-02
---
Added
- Include well properties in outs
- `extend()` method to WellGroup

Changed
- Allow single Well reading for Absorbance, Fluorescence and Luminescence
- Autopick now conforms to updated ASC (**not backwards compatible**)

Removed
- Protocol.plate_to_magblock() and Protocol.plate_from_magblock()

Fixed
- Protocol.stamp() allows one_tip=True when steps use a `mix_vol` greater than "31:microliter" even if transferred volumes are not all greater than "31:microliter"
- Protocol.transfer() respects when `mix_after` or `mix_before` is explicitly False

## v2.5.0 - 2015-10-12
---
Changed
- Protocol.stamp() and Stamp() has been reformatted to take groups of transfers. This allows for one_tip=True, one_source=True, and WellGroup source and destinations

## v2.4.1 - 2015-10-12
---
Fixed
- volume tracking for stamping to/from 384-well plates
- one_tip = True transfers > 750:microliter are transferred with single tip

## v2.4.0 - 2015-09-28
---
Added
- UserError exception class for returning custom errors from within protocol scripts

autoprotocol/harness.py
from __future__ import print_function
import json
import io
from .protocol import Protocol
from .unit import Unit, UnitError
from .container import WellGroup, SEAL_TYPES, COVER_TYPES  # NOQA
from . import UserError
import argparse
import sys

if sys.version_info[0] >= 3:
    string_type = str
else:
    string_type = basestring

'''
    :copyright: 2016 by The Autoprotocol Development Team, see AUTHORS
        for more details.
    :license: BSD, see LICENSE for more details

'''


def param_default(typeDesc):
    if isinstance(typeDesc, string_type):
        typeDesc = {'type': typeDesc}

    type = typeDesc['type']
    default = typeDesc.get('default')

    if default is not None and type != 'group-choice':
        return default

    if typeDesc['type'] in ['aliquot+', 'aliquot++', 'container+']:
        return []
    elif typeDesc['type'] == 'group+':
        return [{}]
    elif typeDesc['type'] == 'group':
        return {
            k: param_default(v)
            for k, v in typeDesc['inputs'].items()
        }
    elif typeDesc['type'] == 'group-choice':
        default_inputs = {}
        for option in typeDesc.get('options', []):
            value = option.get('value')
            inputs = option.get('inputs')
            if inputs:
                group_typedesc = {'type': 'group', 'inputs': inputs}
                default_inputs[value] = param_default(group_typedesc)

        return {
            'value': default,
            'inputs': default_inputs
        }
    elif typeDesc['type'] == 'csv-table':
        return [{}, {}]
    else:
        return None


def convert_param(protocol, val, typeDesc):
    """
    Convert parameters based on their input types

    Parameters
    ----------
    protocol : Protocol
        Protocol object being parsed.
    val : str, int, bool, dict, list
        Parameter value to be converted.
    typeDesc : dict, str
        Description of input type.

    """
    if isinstance(typeDesc, string_type):
        typeDesc = {'type': typeDesc}
    if val is None:
        val = param_default(typeDesc)
    if val is None:  # still None?
        return None
    type = typeDesc['type']

    if type == 'aliquot':
        try:
            container = ('/').join(val.split('/')[0:-1])
            well_idx = val.split('/')[-1]
            return protocol.refs[container].container.well(well_idx)
        except (KeyError, AttributeError, ValueError):
            label = typeDesc.get('label') or "[unknown]"
            raise RuntimeError("'%s' (supplied to input '%s') is not a valid "
                               "reference to an aliquot" % (val, label))
    elif type == 'aliquot+':
        try:
            return WellGroup([convert_param(protocol, a, 'aliquot') for a in val])
        except:
            label = typeDesc.get('label') or "[unknown]"
            raise RuntimeError("The value supplied to input '%s' (type aliquot+) is improperly "
                               "formatted." % label)
    elif type == 'aliquot++':
        try:
            return [convert_param(protocol, aqs, 'aliquot+') for aqs in val]
        except:
            label = typeDesc.get('label') or "[unknown]"
            raise RuntimeError("The value supplied to input '%s' (type aliquot++) is improperly "
                               "formatted." % label)
    elif type == 'container':

        try:
            return protocol.refs[val].container
        except KeyError:
            label = typeDesc.get('label') or "[unknown]"
            raise RuntimeError("'%s' (supplied to input '%s') is not a valid "
                               "reference to a container" % (val, label))
    elif type == 'container+':
        try:
            return [convert_param(protocol, cont, 'container') for cont in val]
        except:
            label = typeDesc.get('label') or "[unknown]"
            raise RuntimeError("The value supplied to input '%s' (type container+) is improperly "
                               "formatted." % label)
    elif type in ['volume', 'time', 'length']:
        try:
            return Unit.fromstring(val)
        except UnitError as e:
            raise RuntimeError("The value supplied (%s) as a unit of '%s' is "
                               "improperly formatted. Units of %s must be in "
                               "the form: 'number:unit'" % (e.value, type, type))
    elif type == 'temperature':
        try:
            if val in ['ambient', 'warm_30', 'warm_37', 'cold_4', 'cold_20', 'cold_80']:
                return val
            else:
                return Unit.fromstring(val)
        except UnitError as e:
            raise RuntimeError("Invalid temperature value for %s: temperature "
                               "input types must be either storage conditions "
                               "(ex: 'cold_20') or temperature units in the "
                               "form of 'number:unit'" % e.value)
    elif type in 'bool':
        return bool(val)
    elif type in 'csv':
        return val
    elif type in ['string', 'choice']:
        return str(val)
    elif type == 'integer':
        try:
            return int(val)
        except ValueError:
            label = typeDesc.get('label') or "[unknown]"
            raise RuntimeError("The value supplied to input '%s' (type integer) is improperly "
                               "formatted." % label)
    elif type == 'decimal':
        try:
            return float(val)
        except ValueError:
            label = typeDesc.get('label') or "[unknown]"
            raise RuntimeError("The value supplied to input '%s' (type decimal) is improperly "
                               "formatted." % label)
    elif type == 'group':
        try:
            return {
                k: convert_param(protocol, val.get(k), typeDesc['inputs'][k])
                for k in typeDesc['inputs']
            }
        except KeyError as e:
            label = typeDesc.get('label') or "[unknown]"
            raise RuntimeError("The value supplied to input '%s' (type group) is missing "
                               "a(n) %s field." % (label, e))
        except AttributeError:
            label = typeDesc.get('label') or "[unknown]"
            raise RuntimeError("The value supplied to input '%s' (type group) is improperly "
                               "formatted." % label)
    elif type == 'group+':
        try:
            return [{
                    k: convert_param(protocol, x.get(k), typeDesc['inputs'][k])
                    for k in typeDesc['inputs']
                    } for x in val]
        except (TypeError, AttributeError):
            raise RuntimeError("The value supplied to input '%s' (type group+) must be in "
                               "the form of a list of dictionaries" % typeDesc['label'])
        except KeyError as e:
            label = typeDesc.get('label') or "[unknown]"
            raise RuntimeError("The value supplied to input '%s' (type group+) is missing "
                               "a(n) %s field." % (label, e))
    elif type == 'group-choice':
        try:
            return {
                'value': val['value'],
                'inputs': {
                    opt['value']: convert_param(
                        protocol,
                        val['inputs'].get(opt['value']),
                        {'type': 'group', 'inputs': opt['inputs']})
                    for opt in typeDesc['options'] if opt['value'] == val['value']
                }
            }
        except (KeyError, AttributeError) as e:
            label = typeDesc.get('label') or "[unknown]"
            if e in ["value", "inputs"]:
                raise RuntimeError("The value supplied to input '%s' (type group-choice) "
                                   "is missing a(n) %s field." % (label, e))
    elif type == 'thermocycle':
        try:
            return [
                {
                    'cycles': g['cycles'],
                    'steps': [convert_param(protocol, s, 'thermocycle_step') for s in g['steps']]
                }
                for g in val
            ]
        except (TypeError, KeyError):
            raise RuntimeError(_thermocycle_error_text())

    elif type == 'thermocycle_step':
        try:
            output = {'duration': Unit.fromstring(val['duration'])}
        except UnitError as e:
            raise RuntimeError("Invalid duration value for %s: duration input types must "
                               "be time units in the form of 'number:unit'" % e.value)

        try:
            if 'gradient' in val:
                output['gradient'] = {
                    'top': Unit.fromstring(val['gradient']['top']),
                    'bottom': Unit.fromstring(val['gradient']['bottom'])
                }
            else:
                output['temperature'] = Unit.fromstring(val['temperature'])
        except UnitError as e:
            raise RuntimeError("Invalid temperature value for %s: thermocycle temperature "
                               "input types must be temperature units in the form of "
                               "'number:unit'" % e.value)

        if 'read' in val:
            output['read'] = val['read']

        return output

    elif type == 'csv-table':
        try:
            values = []
            for i, row in enumerate(val[1]):
                value = {}
                for header, header_value in row.items():
                    typeDesc = {
                        "type": val[0].get(header),
                        "label": "csv-table item (%s): %s" % (i, header)
                    }

                    value[header] = convert_param(protocol, header_value, typeDesc=typeDesc)

                values.append(value)

            return values

        except (AttributeError, IndexError, TypeError):
            label = typeDesc.get('label') or "[unknown]"
            raise RuntimeError(
                "The values supplied to %s (type csv-table) are improperly "
                "formatted. Format must be a list of dictionaries with the first "
                "dictionary comprising keys with associated column input types." % label
            )

    else:
        raise ValueError("Unknown input type %r" % type)


class ProtocolInfo(object):

    def __init__(self, json):
        self.input_types = json['inputs']

    def parse(self, protocol, inputs):
        refs = inputs['refs']
        params = inputs['parameters']

        for name in refs:
            ref = refs[name]
            c = protocol.ref(
                name,
                ref.get('id'),
                ref['type'],
                storage=ref.get('store'),
                discard=ref.get('discard'),
                cover=ref.get('cover'))
            aqs = ref.get('aliquots')
            if aqs:
                for idx in aqs:
                    aq = aqs[idx]
                    c.well(idx).set_volume(aq['volume'])
                    if "name" in aq:
                        c.well(idx).set_name(aq['name'])
                    if "properties" in aq:
                        c.well(idx).set_properties(aq.get('properties'))

        out_params = {}
        for k in self.input_types:
            typeDesc = self.input_types[k]
            out_params[k] = convert_param(protocol, params.get(k), typeDesc)

        return out_params


class Manifest(object):

    """
    Object representation of a manifest.json file

    Parameters
    ----------
    object : JSON object
        A manifest.json file with the following format:

        .. code-block:: json

            {
              "format": "python",
              "license": "MIT",
              "description": "This is a protocol.",
              "protocols": [
                {
                  "name": "SampleProtocol",
                  "version": 1.0.0,
                  "command_string": "python sample_protocol.py",
                  "preview": {
                    "refs":{},
                    "parameters": {},
                  "inputs": {},
                  "dependencies": []
                }
              ]
            }

    """

    def __init__(self, json):
        '''
        '''
        self.protocols = json['protocols']

    def protocol_info(self, name):
        '''
        '''
        try:
            return ProtocolInfo(
                next(p for p in self.protocols if p['name'] == name))
        except StopIteration:
            raise RuntimeError("Harness.run(): %s does not match "
                               "the 'name' field of any protocol in the "
                               "associated manifest.json file." % name)


def run(fn, protocol_name=None, seal_after_run=True):
    """
    Run the protocol specified by the function.

    If protocol_name is passed, use preview parameters from the protocol with
    the matching "name" value in the manifest.json file to run the given
    function.  Otherwise, take configuration JSON file from the command line
    and run the given function.

    Parameters
    ----------
    fn : function
        Function that generates Autoprotocol
    protocol_name :  str, optional
        str matching the "name" value in the manifest.json file
    seal_after_run : bool, optional
        Implicitly add a seal/cover to all stored refs within the protocol
        using seal_on_store()
    """
    parser = argparse.ArgumentParser()
    parser.add_argument(
        'config',
        help='JSON-formatted protocol configuration file')
    args = parser.parse_args()

    source = json.loads(io.open(args.config, encoding='utf-8').read())
    protocol = Protocol()
    if protocol_name:
        manifest_json = io.open('manifest.json', encoding='utf-8').read()
        manifest = Manifest(json.loads(manifest_json))
        params = manifest.protocol_info(protocol_name).parse(protocol, source)
    else:
        params = protocol._ref_containers_and_wells(source["parameters"])

    try:
        fn(protocol, params)
        if seal_after_run:
            seal_on_store(protocol)
    except UserError as e:
        print(json.dumps({
            'errors': [
                {
                    'message': e.message,
                    'info': e.info
                }
            ]
        }, indent=2))
        return

    print(json.dumps(protocol.as_dict(), indent=2))


def _thermocycle_error_text():
    """
    Returns formatted error text for thermocycle value errors
    """

    return """Thermocycle input types must take a list of dictionaries in the form of:
  [{"cycles": integer,
    "steps": [{
      "duration": duration,
      "temperature": temperature
      "read": boolean (optional)
    }]
  }]
--or--
  [{"cycles": integer,
    "steps": [{
      "duration": duration,
      "gradient": {
        "top": temperature,
        "bottom": temperature
      }
      "read": boolean (optional)
    }]
  }]
(You can intermix gradient and non-gradient steps)"""


def seal_on_store(protocol):
    '''
    Implicitly adds seal/cover instructions to the end of a run for containers
    that do not have a cover.   Cover type applied defaults first to
    "seal" if its within the capabilities of the container type, otherwise
    to "cover".

    Example Usage:

        .. code-block:: python

            def example_method(protocol, params):
            cont = params['container']
            p.transfer(cont.well("A1"), cont.well("A2"), "10:microliter")
            p.seal(cont)
            p.unseal(cont)
            p.cover(cont)
            p.uncover(cont)

    Autoprotocol Output:

        .. code-block:: json

            {
              "refs": {
                "plate": {
                  "new": "96-pcr",
                  "store": {
                    "where": "ambient"
                  }
                }
              },
              "instructions": [
                {
                  "groups": [
                    {
                      "transfer": [
                        {
                          "volume": "10.0:microliter",
                          "to": "plate/1",
                          "from": "plate/0"
                        }
                      ]
                    }
                  ],
                  "op": "pipette"
                },
                {
                  "object": "plate",
                  "type": "ultra-clear",
                  "op": "seal"
                },
                {
                  "object": "plate",
                  "op": "unseal"
                },
                {
                  "lid": "universal",
                  "object": "plate",
                  "op": "cover"
                },
                {
                  "object": "plate",
                  "op": "uncover"
                },
                {
                  "type": "ultra-clear",
                  "object": "plate",
                  "op": "seal"
                }
              ]
            }

    '''
    for name, ref in protocol.refs.items():
        if "store" in ref.opts.keys():
            if not (ref.container.is_covered() or ref.container.is_sealed()):
                if "seal" in ref.container.container_type.capabilities:
                    protocol.seal(ref.container,
                                  ref.container.container_type.seal_types[0])
                elif "cover" in ref.container.container_type.capabilities:
                    protocol.cover(ref.container,
                                   ref.container.container_type.cover_types[0])
                else:
                    continue
